# Explore and cluster the neighborhoods in Toronto Part - 1 To 3
### Scrape Wikipedia page https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M


# Option A
### We can use BeautifulSoup


# Option B 
### Use python pandas 

### visit the web page "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
##### Explore the web page
##### It looks not very complicated so python pandas can be used to read & load the data

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis

print('Libraries imported.')

Libraries imported.


In [2]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

wikipage = pd.read_html(url)



# Explore Data and its struture

In [3]:
type(wikipage)

list

In [4]:
len(wikipage)

3

### Check the every element of list 'wikipage[0], 'wikipage[1], 'wikipage[2] one by one

In [5]:
wikipage[0].head()

0             1                 2
0  Postcode       Borough     Neighbourhood
1       M1A  Not assigned      Not assigned
2       M2A  Not assigned      Not assigned
3       M3A    North York         Parkwoods
4       M4A    North York  Victoria Village

In [6]:
wikipage[1].head()

0                      1   \
0                                                NaN  Canadian postal codes   
1  NL NS PE NB QC ON MB SK AB BC NU/NT YT A B C E...                     NL   
2                                                 NL                     NS   
3                                                  A                      B   

    2    3    4    5    6    7    8    9  ...    21   22   23   24   25   26  \
0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN ...   NaN  NaN  NaN  NaN  NaN  NaN   
1   NS   PE   NB   QC   ON   MB   SK   AB ...     L    M    N    P    R    S   
2   PE   NB   QC   ON   MB   SK   AB   BC ...   NaN  NaN  NaN  NaN  NaN  NaN   
3    C    E    G    H    J    K    L    M ...   NaN  NaN  NaN  NaN  NaN  NaN   

    27   28   29   30  
0  NaN  NaN  NaN  NaN  
1    T    V    X    Y  
2  NaN  NaN  NaN  NaN  
3  NaN  NaN  NaN  NaN  

[4 rows x 31 columns]

In [7]:
wikipage[2].head()


0   1   2   3   4   5   6   7   8   9      10  11   12   13   14   15   16  \
0  NL  NS  PE  NB  QC  ON  MB  SK  AB  BC  NU/NT  YT  NaN  NaN  NaN  NaN  NaN   
1   A   B   C   E   G   H   J   K   L   M      N   P    R    S    T    V    X   

    17  
0  NaN  
1    Y

# Data Selection

### Now it is clear wikipage is a nested list structure and our data is in "wikipage[0]"
#### Create a data frame Totonto_df with wikipage[0]

In [8]:
Totonto_df = wikipage[0]
Totonto_df.head()

0             1                 2
0  Postcode       Borough     Neighbourhood
1       M1A  Not assigned      Not assigned
2       M2A  Not assigned      Not assigned
3       M3A    North York         Parkwoods
4       M4A    North York  Victoria Village

### First row is holds the name of columns 
#### so we will make it the column name of daraframe  and delete the first row

In [9]:
Totonto_df[0:1].values.tolist()

[['Postcode', 'Borough', 'Neighbourhood']]

In [10]:
Totonto_df.columns=['Postcode','Borough','Neighborhood']

#### Remove first row it holds the name of column and reset the index

In [11]:
#Totonto_df = Totonto_df.iloc[1:]
Totonto_df.drop(Totonto_df.index[:1], inplace=True)
Totonto_df.reset_index(drop=True, inplace=True)
Totonto_df.head()

Postcode           Borough      Neighborhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront

In [12]:
Totonto_df.shape

(289, 3)

# Data Preparation


### Target 1. The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood

In [13]:
Totonto_df.head()

Postcode           Borough      Neighborhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront

### Target 2. Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.


In [14]:
Totonto_df.replace({'Borough': 'Not assigned' }, np.nan, inplace = True)
Totonto_df.head()
# simply drop whole row with NaN in "price" column
Totonto_df.dropna(subset=["Borough"], axis=0, inplace=True)

# reset index, because we droped two rows
Totonto_df.reset_index(drop=True, inplace=True)
Totonto_df.head()

Postcode           Borough      Neighborhood
0      M3A        North York         Parkwoods
1      M4A        North York  Victoria Village
2      M5A  Downtown Toronto      Harbourfront
3      M5A  Downtown Toronto       Regent Park
4      M6A        North York  Lawrence Heights

In [15]:
Totonto_df.shape

(212, 3)

### Target 3. More than one neighborhood can exist in one postal code area. 
#### For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11 in the above table.

In [16]:
# Groupby and join can be used for the purpose
Totonto_df = Totonto_df.groupby(['Postcode','Borough'])['Neighborhood'].apply(', '.join).reset_index()
Totonto_df.head()

Postcode      Borough                            Neighborhood
0      M1B  Scarborough                          Rouge, Malvern
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union
2      M1E  Scarborough       Guildwood, Morningside, West Hill
3      M1G  Scarborough                                  Woburn
4      M1H  Scarborough                               Cedarbrae

In [17]:
Totonto_df.shape

(103, 3)

### Target 4. If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough. So for the 9th cell in the table on the Wikipedia page, the value of the Borough and the Neighborhood columns will be Queen's Park.


In [18]:
#Totonto_df[ (Totonto_df.Neighbourhood=='Not assigned')]['Neighborhood'] = [Borough]
Totonto_df['Neighborhood'] = np.where(Totonto_df['Neighborhood'] == 'Not assigned', Totonto_df['Borough'], Totonto_df['Neighborhood'])
Totonto_df.head()

Postcode      Borough                            Neighborhood
0      M1B  Scarborough                          Rouge, Malvern
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union
2      M1E  Scarborough       Guildwood, Morningside, West Hill
3      M1G  Scarborough                                  Woburn
4      M1H  Scarborough                               Cedarbrae

### Target 5. Use the .shape method to print the number of rows of your dataframe.

In [19]:
Totonto_df.shape

(103, 3)

# Explore and cluster the neighborhoods in Toronto Part - 2

In [20]:
!conda install -c conda-forge geopy --yes # geopy
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

print('Libraries imported.')

Fetching package metadata .............
Solving package specifications: .

Package plan for installation in environment /opt/conda/envs/DSX-Python35:

The following NEW packages will be INSTALLED:

    geographiclib: 1.49-py_0   conda-forge
    geopy:         1.18.1-py_0 conda-forge

geographiclib- 100% |################################| Time: 0:00:00   3.59 MB/s
geopy-1.18.1-p 100% |################################| Time: 0:00:00   3.96 MB/s
Libraries imported.


## Target - Use the Geocoder package or the csv file to create the dataframe:
#### Postcode, Borough, Neighbourhood, Latitude, Longitude
#### Option 1 - Geocoder
#### Option 2 - csv File

## Option 1(A) - Geocoder

# Issues - 
#### 1. Post code is assigned to more then one Neighborhood
#### 2. Borough to more then one post code
#### both above condition adds complexity

## Option 1(B) - Geopy - Nominatim

In [21]:
to_post_ll_df = pd.DataFrame(columns=['Postcode', 'Latitude', 'Longitude'])
bro = ''
for index, row in Totonto_df.iterrows():
    geolocator = Nominatim(user_agent="ny_explorer")
    location = geolocator.geocode(row['Borough'] + ' Canada')
    latitude = location.latitude
    longitude = location.longitude
    #print('The geograpical coordinate of  {} are {}, {}.'.format(row['Postcode'],latitude, longitude))
    to_post_ll_df=to_post_ll_df.append([{'Postcode':row['Postcode'],'Latitude':latitude,'Longitude':longitude}], ignore_index=True)
to_post_ll_df.head()

Postcode   Latitude  Longitude
0      M1B  43.773077 -79.257774
1      M1C  43.773077 -79.257774
2      M1E  43.773077 -79.257774
3      M1G  43.773077 -79.257774
4      M1H  43.773077 -79.257774

## Option 2 - csv File
### 1. DownLoad data 

In [22]:
path = 'https://cocl.us/Geospatial_data'
g_df = pd.read_csv(path)
g_df.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

## 2. Merage dataframe on Postcode
### 1. Rename the column 'Postal Code' To 'Postcode'
#### 2. Merge Totonto_df with g_df on 'PostCode'
#### 3. (how = 'inner' is used )

In [23]:
g_df = g_df.rename(columns={'Postal Code': 'Postcode'})
Totonto_Geo_df = pd.merge(Totonto_df, g_df, how='inner', on=['Postcode'])

In [24]:
Totonto_Geo_df.shape

(103, 5)

In [25]:
Totonto_Geo_df.head()

Postcode      Borough                            Neighborhood   Latitude  \
0      M1B  Scarborough                          Rouge, Malvern  43.806686   
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2      M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3      M1G  Scarborough                                  Woburn  43.770992   
4      M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

# Explore and cluster the neighborhoods in Toronto Part - 3

### Issue - IBM Watson Notebook has problem in rendering 'folium.CircleMarker'

### "folium.CircleMarker" in this notebook work good with cognitiveclass jupyterlab  

##### Import Libraries

In [26]:
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

!conda install -c conda-forge folium=0.5.0 --yes 
import folium # map rendering library

print('Libraries imported.')

Fetching package metadata .............
Solving package specifications: .

Package plan for installation in environment /opt/conda/envs/DSX-Python35:

The following NEW packages will be INSTALLED:

    altair:  2.2.2-py35_1 conda-forge
    branca:  0.3.1-py_0   conda-forge
    folium:  0.5.0-py_0   conda-forge
    vincent: 0.4.4-py_1   conda-forge

altair-2.2.2-p 100% |################################| Time: 0:00:00   6.72 MB/s
branca-0.3.1-p 100% |################################| Time: 0:00:00  33.14 MB/s
vincent-0.4.4- 100% |################################| Time: 0:00:00  39.49 MB/s
folium-0.5.0-p 100% |################################| Time: 0:00:00  44.73 MB/s
Libraries imported.


###  1. Only boroughs that contain the word Toronto (Cluster)

In [27]:
Totonto_df=Totonto_Geo_df[Totonto_Geo_df['Borough'].str.contains('Toronto')]
Totonto_df.head()

Postcode          Borough                    Neighborhood   Latitude  \
37      M4E     East Toronto                     The Beaches  43.676357   
41      M4K     East Toronto    The Danforth West, Riverdale  43.679557   
42      M4L     East Toronto  The Beaches West, India Bazaar  43.668999   
43      M4M     East Toronto                 Studio District  43.659526   
44      M4N  Central Toronto                   Lawrence Park  43.728020   

    Longitude  
37 -79.293031  
41 -79.352188  
42 -79.315572  
43 -79.340923  
44 -79.388790

### Get the latitude and longitude of Totonto

In [28]:
city= 'Toronto, Canada'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(city)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Canada are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Canada are 43.653963, -79.387207.


### Create Map with - 
#### 1. Create map of Toronto's neighborhood, 
#### 2. Postal code M  
#### 3. Borough contains 'Toronto' using latitude and longitude values

In [29]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
folium.CircleMarker(
    [latitude, longitude],
    radius=10,
    popup='Canada',
    color='red',
    fill=True,
    fill_color='red',
    fill_opacity=0.6,
    parse_html=False).add_to(map_toronto) 

  # add markers to map
for lat, lng, borough, neighborhood in zip(Totonto_df['Latitude'], Totonto_df['Longitude'], Totonto_df['Borough'], Totonto_df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)
    
map_toronto

### Issue - IBM Watson Notebook has problem in rendering 'folium.CircleMarker'
### "folium.CircleMarker" in this notebook work good with cognitiveclass jupyterlab

#### Dataframe of Canada with postcode M

In [31]:
canadaPcodeM_df = Totonto_Geo_df 

In [32]:

print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(canadaPcodeM_df['Borough'].unique()),
        canadaPcodeM_df.shape[0]
    )
)
boroughL = canadaPcodeM_df['Borough'].unique().tolist()
boroughL

The dataframe has 11 boroughs and 103 neighborhoods.


['Scarborough',
 'North York',
 'East York',
 'East Toronto',
 'Central Toronto',
 'Downtown Toronto',
 'York',
 'West Toronto',
 "Queen's Park",
 'Mississauga',
 'Etobicoke']

### Get random color for  dataframe has 11 boroughs with 103 neighborhoods.


In [33]:
import random

number_of_colors = 11

ncolor = ["#"+''.join([random.choice('0123456789ABCDEF') for j in range(6)])
             for i in range(number_of_colors)]
print(ncolor)

['#CA67F3', '#4BD769', '#FF76EB', '#A53C0A', '#3519F2', '#D99BD5', '#687409', '#D67262', '#48D094', '#2CE0E4', '#871112']


## Map all 11 borough and their neighborhoods with random color


In [34]:
city= 'Toronto, Canada'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(city)
latitude = location.latitude
longitude = location.longitude

map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
folium.CircleMarker(
    [latitude, longitude],
    radius=10,
     popup='Canada',
    color='red',
    fill=True,
    fill_color='red',
    fill_opacity=0.6,
    parse_html=False).add_to(map_toronto) 
i=0 # for color Selection
for br in boroughL: # for each boroughs in list 
    
    boroughNbr_df = canadaPcodeM_df[canadaPcodeM_df['Borough'].str.match(br)]
      # add markers to map
    for lat, lng, borough, neighborhood in zip(boroughNbr_df['Latitude'], boroughNbr_df['Longitude'], boroughNbr_df['Borough'], boroughNbr_df['Neighborhood']):
        label = '{}, {}'.format(neighborhood, borough)
        label = folium.Popup(label, parse_html=True)
        folium.CircleMarker(
            [lat, lng],
            radius=5,
            popup=label,
            color=ncolor[i],
            fill=True,
            fill_color=ncolor[i],
            fill_opacity=0.7,
            parse_html=False).add_to(map_toronto)
    i=i+1
map_toronto

### Issue - IBM Watson Notebook has problem in rendering 'folium.CircleMarker'
### "folium.CircleMarker" in this notebook work good with cognitiveclass jupyterlab